pkg/offline/offline.go

```diff
@@ -157,6 +157,146 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
 	}, nil
 }
 
+// Sync returns a function to send queued heartbeats to the WakaTime API.
+func Sync(filepath string, syncLimit int) func(next heartbeat.Handle) error {
+	return func(next heartbeat.Handle) error {
+		log.Debugf("execute offline sync with file %s", filepath)
+
+		hh, err := popHeartbeats(filepath, syncLimit)
+		if err != nil {
+			return fmt.Errorf("failed to fetch heartbeat from offline queue: %s", err)
+		}
+
+		if len(hh) == 0 {
+			log.Debugln("abort execution, as there are no queued heartbeats ready for sending")
+
+			return nil
+		}
+
+		results, err := next(hh)
+		if err != nil {
+			requeueErr := pushHeartbeats(filepath, hh)
+			if requeueErr != nil {
+				log.Errorf("failed to push heatbeats after api error to queue: %s", requeueErr)
+			}
+
+			return err
+		}
+
+		return handleResults(filepath, results, hh)
+	}
+}
+
+func handleResults(filepath string, results []heartbeat.Result, hh []heartbeat.Heartbeat) error {
+	var (
+		err               error
+		withInvalidStatus []heartbeat.Heartbeat
+	)
+
+	// push heartbeats with invalid result status codes to queue
+	for n, result := range results {
+		if n >= len(hh) {
+			log.Warnln("results from api not matching heartbeats sent")
+			break
+		}
+
+		if result.Status != http.StatusCreated &&
+			result.Status != http.StatusAccepted &&
+			result.Status != http.StatusBadRequest {
+			withInvalidStatus = append(withInvalidStatus, hh[n])
+		}
+	}
+
+	if len(withInvalidStatus) > 0 {
+		log.Debugf("pushing %d heartbeat(s) with invalid result to queue", len(withInvalidStatus))
+
+		err = pushHeartbeats(filepath, withInvalidStatus)
+		if err != nil {
+			log.Errorf("failed to push heatbeats with invalid status to queue: %s", err)
+		}
+	}
+
+	// handle leftover heartbeats
+	leftovers := len(hh) - len(results)
+	if leftovers > 0 {
+		log.Warnf("Missing %d results from api.", leftovers)
+
+		start := len(hh) - leftovers
+
+		err = pushHeartbeats(filepath, hh[start:])
+		if err != nil {
+			log.Errorf("failed to push leftover heatbeats to queue: %s", err)
+		}
+	}
+
+	return err
+}
+
+func popHeartbeats(filepath string, syncLimit int) ([]heartbeat.Heartbeat, error) {
```

Comment:
`syncLimit` should never be larger than `24`, but it looks like this comes directly from user input?

Reply:
`syncLimit` should just be `limit` because this code has nothing to do with syncing.. it's just popping multiple (up to 24) heartbeats from the offline db.

Reply:
Alright, makes sense :+1: So I guess we should silently use 24, in case the user provides any higher value. Additionally a log message on warning level to make the user aware of this.

Reply:
Default should be 1k.

Reply:
Alright, will adjust this in https://github.com/wakatime/wakatime-cli/pull/455